### PR TITLE
fix: export pivot results on csv

### PR DESCRIPTION
**Pull request by rephus**
Closes:  https://github.com/lightdash/lightdash/issues/6838

### Description:

- [x] supports custom labels
- [x] supports hidden fields
- [x] Download csv from dashboard 
- [x] monitor method using sentry transaction
- [x] Raw/formatted
- [x] multi pivot
- [x] Explorer
- [x]  Chart
- [x]  Scheduled deliveries
- [x] analytic events  : added `numPivotDimensions` to `download_results.completed`

### Not done in this PR 

Export from: (some might work if they use the same method) 
- [ ]  SQL runner
- [ ] metrics as rows

**single pivot**

Screenshot from 2024-11-11 10-33-59

**multi pivot**
Screenshot from 2024-11-11 10-34-20

**raw format**
Screenshot from 2024-11-11 10-39-20

**custom labels**
Screenshot from 2024-11-11 10-49-19

** Explore & no index **

Screenshot from 2024-11-12 10-44-31

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

**Comment by rephus** [comment]
Context, after implementing this, we realized we want to do this on embedding first, so we're shipping this as it is

File: packages/common/src/pivotTable/pivotQueryResults.ts [hunk 1]
```diff
@@ -1,19 +1,16 @@
+import isNumber from 'lodash/isNumber';
```
Comment (rephus) [inline packages/common/src/pivotTable/pivotQueryResults.ts:1]: I didn't change anything in this file, it must have been a formatting issue

**Comment by rephus** [comment]
@IrakliJani this was implemented before the export on embedding, so I probably need to apply the same fixes we did for the edge cases you found.

File: packages/common/src/pivotTable/pivotQueryResults.ts [hunk 2]
```diff
@@ -767,6 +761,68 @@ export const pivotQueryResults = ({
             pivotColumnInfo: [],
         },
     };
-
     return combinedRetrofit(pivotData, getField, getFieldLabel);
 };
+
+export const pivotResultsAsCsv = (
```
Comment (rephus) [inline packages/common/src/pivotTable/pivotQueryResults.ts:767]: I created this method here, so we can also use it in the frontend for exporting on embedding (similar api)

File: packages/frontend/src/components/Explorer/VisualizationCard/VisualizationCard.tsx [hunk 3]
```diff
@@ -127,6 +129,7 @@ const VisualizationCard: FC<{
                 hiddenFields: getHiddenTableFields(
                     unsavedChartVersion.chartConfig,
                 ),
+                pivotColumns: pivotConfig?.columns,
```
Comment (rephus) [inline packages/frontend/src/components/Explorer/VisualizationCard/VisualizationCard.tsx:132]: This is for exporting from explore and saved charts

File: packages/frontend/src/hooks/useColumnTotals.ts [hunk 4]
```diff
@@ -20,27 +15,6 @@ type Args = {
     itemsMap: Record<FieldId, Field | TableCalculation>;
 };
 
-export const isSummable = (item: Item | undefined) => {
```
Comment (rephus) [inline packages/frontend/src/hooks/useColumnTotals.ts:23]: used for pivotResults, so moved to common

File: packages/common/src/pivotTable/pivotQueryResults.ts (hunk 1 above)
Comment (rephus) [inline packages/common/src/pivotTable/pivotQueryResults.ts:1]: Moved from the frontend, to common, so we can use it both frontend adn backend

File: packages/common/package.json [hunk 5]
```diff
@@ -20,6 +20,8 @@
         "cronstrue": "^2.23.0",
         "dayjs": "^1.11.9",
         "dependency-graph": "^0.11.0",
+        "type-fest": "^3.10.0",
```
Comment (rephus) [inline packages/common/package.json:23]: required by pivot results

File: packages/common/src/pivotTable/pivotQueryResults.ts (hunk 2 above)
Comment (rephus) [reply packages/common/src/pivotTable/pivotQueryResults.ts:767]: This code is simila to what you reviewed yesterday fro embedding (I copied it from there to ensure we cover all the edge cases we found yesterday)

**Comment by rephus** [comment]
FYI @TuringLovesDeathMetal I added a new property to analytic events 

 analytic events : added `numPivotDimensions: number` to `download_results.completed` 

if > 0, then we're exporting a pivoted csv file.

**Comment by rephus** [comment]
I need to fix tests

File: packages/frontend/src/components/Explorer/ResultsCard/ResultsCard.tsx [hunk 6]
```diff
@@ -64,6 +64,7 @@ const ResultsCard: FC = memo(() => {
             onlyRaw,
             columnOrder,
             showTableNames: true,
+            pivotColumns: undefined, // results are always unpivoted
```
Comment (IrakliJani) [inline packages/frontend/src/components/Explorer/ResultsCard/ResultsCard.tsx:67]: 👍

File: packages/frontend/src/hooks/tableVisualization/useTableConfig.ts [hunk 7]
```diff
@@ -23,11 +24,10 @@ import {
     type TableHeader,
 } from '../../components/common/Table/types';
 import { useCalculateTotal } from '../useCalculateTotal';
-import { isSummable } from '../useColumnTotals';
 import getDataAndColumns from './getDataAndColumns';
 
 const createWorker = createWorkerFactory(
-    () => import('../pivotTable/pivotQueryResults'),
+    () => import('@lightdash/common/src/pivotTable/pivotQueryResults'),
```
Comment (IrakliJani) [inline packages/frontend/src/hooks/tableVisualization/useTableConfig.ts:30]: 👍

File: packages/common/src/pivotTable/pivotQueryResults.ts [hunk 8]
```diff
@@ -41,19 +38,17 @@ type RecursiveRecord<T = unknown> = {
     [key: string]: RecursiveRecord<T> | T;
 };
 
-const isRecursiveRecord = (value: unknown): value is RecursiveRecord => {
-    return typeof value === 'object' && value !== null;
-};
+const isRecursiveRecord = (value: unknown): value is RecursiveRecord =>
```
Comment (IrakliJani) [inline packages/common/src/pivotTable/pivotQueryResults.ts:41]: I don't know about you but I find this rule quite annoying

**Review by IrakliJani — COMMENTED** [review]
changes looks great! amazing work 🎉 I'll test functionalities and report if I find any bugs

File: packages/common/src/pivotTable/pivotQueryResults.ts (hunk 8 above)
Comment (rephus) [reply packages/common/src/pivotTable/pivotQueryResults.ts:41]: I did'nt change anything on this functions, they were simply copied over. I prefer not to change the behaviour at the moment, unless you have a better suggestion

**Review by IrakliJani — APPROVED** [review]

File: packages/common/src/pivotTable/pivotQueryResults.ts (hunk 8 above)
Comment (IrakliJani) [reply packages/common/src/pivotTable/pivotQueryResults.ts:41]: all good! it's good to know that it's just not me 😊

**Pull request merged**